File: panel/modules/renders/ip.js
// panel/modules/renders/ip.js
import { field, fmtDate, escapeHtml } from "../utils.js";

export function renderIpResult(r) {
    const parts = [];

    const g = r.geo || {};
    if (Object.keys(g).length) {
        const coords = (g.lat && g.lon) ? `${g.lat}, ${g.lon}` : null;
        const mapLink = (g.lat && g.lon)
            ? `<a href="https://www.openstreetmap.org/?mlat=${g.lat}&mlon=${g.lon}#map=12/${g.lat}/${g.lon}" target="_blank" rel="noopener">${coords}</a>`
            : coords;
        parts.push(`
            <details open>
                <summary>Гео</summary>
                <table class="kv">
                    ${field("Страна", g.country ? `${g.country} (${g.country_code || "?"})` : null)}
                    ${field("Регион", g.region)}
                    ${field("Город", g.city)}
                    ${field("Индекс", g.zip)}
                    ${field("Координаты", mapLink)}
                    ${field("Часовой пояс", g.timezone)}
                </table>
            </details>
        `);
    } else if (r.geo_error) {
        parts.push(`<div class="warn">Гео: ${escapeHtml(r.geo_error)}</div>`);
    }

    const n = r.network || {};
    if (Object.keys(n).length) {
        const flags = [];
        if (n.mobile) flags.push("📱 мобильный");
        if (n.proxy) flags.push("🎭 прокси/VPN");
        if (n.hosting) flags.push("🖥 хостинг/датацентр");
        parts.push(`
            <details open>
                <summary>Провайдер</summary>
                <table class="kv">
                    ${field("ISP", n.isp)}
                    ${field("Организация", n.org)}
                    ${field("ASN", n.as)}
                    ${field("ASN name", n.asname)}
                    ${field("Признаки", flags.length ? flags.join(", ") : "—")}
                </table>
            </details>
        `);
    }

    const ptr = r.ptr || [];
    parts.push(`
        <details>
            <summary>Reverse DNS (PTR)</summary>
            ${ptr.length
                ? `<table class="kv">${ptr.map(p => field("PTR", p)).join("")}</table>`
                : `<div class="warn">PTR-записи не найдены</div>`}
        </details>
    `);

    const w = r.whois || {};
    if (Object.keys(w).length) {
        parts.push(`
            <details>
                <summary>Whois (RIR)</summary>
                <table class="kv">
                    ${field("Network", w.netname)}
                    ${field("CIDR", w.cidr)}
                    ${field("Диапазон", w.start_address && w.end_address ? `${w.start_address} — ${w.end_address}` : null)}
                    ${field("Страна", w.country)}
                    ${field("ASN", w.asn)}
                    ${field("ASN описание", w.asn_description)}
                    ${field("ASN страна", w.asn_country)}
                    ${field("Зарегистрирован", fmtDate(w.registered))}
                </table>
            </details>
        `);
    } else if (r.whois_error) {
        parts.push(`<div class="warn">Whois: ${escapeHtml(r.whois_error)}</div>`);
    }

    return parts.join("");
}

File: agent/modules/ip.py
# agent/modules/ip.py
import dns.resolver
import dns.reversename
import requests


def module_ip(target: str) -> dict:
    target = target.strip()
    result = {"target": target}

    # --- 1. Гео/провайдер через ip-api.com ---
    try:
        r = requests.get(
            f"http://ip-api.com/json/{target}",
            params={
                "fields": "status,message,country,countryCode,region,regionName,"
                          "city,zip,lat,lon,timezone,isp,org,as,asname,"
                          "mobile,proxy,hosting,query"
            },
            timeout=10,
        )
        data = r.json()
        if data.get("status") == "success":
            result["geo"] = {
                "country": data.get("country"),
                "country_code": data.get("countryCode"),
                "region": data.get("regionName"),
                "city": data.get("city"),
                "zip": data.get("zip"),
                "lat": data.get("lat"),
                "lon": data.get("lon"),
                "timezone": data.get("timezone"),
            }
            result["network"] = {
                "isp": data.get("isp"),
                "org": data.get("org"),
                "as": data.get("as"),
                "asname": data.get("asname"),
                "mobile": data.get("mobile"),
                "proxy": data.get("proxy"),
                "hosting": data.get("hosting"),
            }
        else:
            result["geo_error"] = data.get("message", "unknown error")
    except Exception as e:
        result["geo_error"] = str(e)

    # --- 2. Reverse DNS (PTR) ---
    try:
        rev = dns.reversename.from_address(target)
        answers = dns.resolver.resolve(rev, "PTR", lifetime=5)
        result["ptr"] = [str(r).rstrip(".") for r in answers]
    except Exception:
        result["ptr"] = []

    # --- 3. Whois по IP через ipwhois ---
    try:
        from ipwhois import IPWhois
        obj = IPWhois(target)
        whois_data = obj.lookup_rdap(depth=1)

        net = whois_data.get("network", {}) or {}
        result["whois"] = {
            "netname": net.get("name"),
            "cidr": net.get("cidr"),
            "country": net.get("country"),
            "start_address": net.get("start_address"),
            "end_address": net.get("end_address"),
            "registered": net.get("events", [{}])[0].get("timestamp") if net.get("events") else None,
            "description": net.get("remarks"),
            "asn": whois_data.get("asn"),
            "asn_description": whois_data.get("asn_description"),
            "asn_country": whois_data.get("asn_country_code"),
        }
    except Exception as e:
        result["whois_error"] = str(e)

    return result